ESQL: Fix MV_DEDUPE when using data from an index

alex-spies opened the pull request:
Fix https://github.com/elastic/elasticsearch/issues/104745.

alex-spies committed:
Do not assume docs values are deduped

alex-spies committed:
Fix bug in CsvAssert

alex-spies committed:
Add test

alex-spies committed:
Align boolean mv dedupe with other types

elasticsearchmachine commented:
Hi @alex-spies, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

alex-spies committed:
Update docs/changelog/107577.yaml

alex-spies commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperator.java:
```diff
@@ -580,7 +580,7 @@ private ComputeBlockLoaderFactory(BlockFactory factory, int pageSize) {
         @Override
         public BlockLoader.BooleanBuilder booleansFromDocValues(int expectedCount) {
-            return factory.newBooleanBlockBuilder(expectedCount).mvOrdering(Block.MvOrdering.DEDUPLICATED_AND_SORTED_ASCENDING);
+            return factory.newBooleanBlockBuilder(expectedCount);
```
@nik9000 it probably had a very good reason why we assumed this MvOrdering, but in practice unordered multi-values with dups, loaded from an index, ended up being marked as deduplicated which mv_dedupe used to optimize itself away into a no-op.

nik9000 replied:
Booleans come from https://lucene.apache.org/core/8_0_0/core/index.html?org/apache/lucene/document/SortedNumericDocValuesField.html which is sorted. They are stored as `0` and `1` and sorted. So all the `false` values should be before all of the `true` values.

So, I think, these should be `Block.MvOrdering.SORTED_ASCENDING`. Which a constant we don't have. FWIW it's much more efficient to deduplicate those. But we can get that code later.

I'm unsure how this made it this long. Numerics are not deduplicated - we can't pretend they are. Nor can we for booleans.

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/string.csv-spec:
```diff
@@ -434,6 +434,16 @@ ROW a=["foo", "foo", "bar", "foo"]
+;
+
+b:boolean     | k:keyword      | i:integer | l:long | d:double
+[false, true] | ["bar", "foo"] | [1, 2]    | [1, 2] | [1.1, 2.2]
```
I have no idea why, but without this fix, the multivalues are _not deduped_, but _indeed sorted_ when running the query from above; so this fails with

```
row 0 column 2:a list containing
row 0 column 2:0: expected "[1, 2]" but was "[1, 1, 2, 2]"
row 0 column 3:a list containing
row 0 column 3:0: expected "[1, 2]" but was "[1, 1, 2, 2]"
row 0 column 4:a list containing
row 0 column 4:0: expected "[1.1, 2.2]" but was "[1.1, 1.1, 2.2, 2.2]"
```

nik9000 replied:
It's because numeric *are* sorted, but not deduplicated. Disaster.

alex-spies committed:
Add skip mark to test

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into fi-mv-dedupe

alex-spies committed:
Update csv test skip for 8.15

alex-spies committed:
Update test

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/MultivalueDedupeBoolean.java:
```diff
@@ -43,7 +43,7 @@ public MultivalueDedupeBoolean(BooleanBlock block) {
      */
     public BooleanBlock dedupeToBlock(BlockFactory blockFactory) {
-        if (false == block.mayHaveMultivaluedFields()) {
+        if (block.mvDeduplicated()) {
```
:+1:

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperator.java:
```diff
@@ -590,7 +590,7 @@ public BlockLoader.BooleanBuilder booleans(int expectedCount) {
         @Override
         public BlockLoader.BytesRefBuilder bytesRefsFromDocValues(int expectedCount) {
-            return factory.newBytesRefBlockBuilder(expectedCount).mvOrdering(Block.MvOrdering.DEDUPLICATED_AND_SORTED_ASCENDING);
+            return factory.newBytesRefBlockBuilder(expectedCount);
```
But these *are* deduplicated. We should keep this line as is.

nik9000 commented:
Thanks for getting this! I'd feel better if we added a new enum constanted for `SORTED_ASCENDING` and stuck it on the numerics in this PR. And kept the current constant on bytes. The new constant won't do anything, but we can plug it in easily enough later.

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into fix-mv-dedupe

alex-spies committed:
Correctly tag doc blocks as SORTED_ASCENDING

Also tag bytes ref blocks from doc values as deduped and sorted.

alex-spies committed:
Update tests

alex-spies commented:
Thanks @nik9000 , added the new constant and updated the orderings in the operator, corresponding tests as well as the doc of the `BlockFactory` interface.

astefan commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/Block.java:
```diff
@@ -121,6 +121,7 @@ public interface Block extends Accountable, BlockLoader.Block, NamedWriteable, R
     enum MvOrdering {
         UNORDERED(false, false),
         DEDUPLICATED_UNORDERD(true, false),
+        SORTED_ASCENDING(false, true),
```
We are serializing the MvOrdering as well and a newer ESQL version reading from an old version Enum I _think_ it will be confused by the position in the Enum of this new value. Take a look at `StreamOutput.writeEnum`.

What I'm trying to say is that, because the enum values are serialized using their values' position in the enum, new values added to a serializable enum should be added at the end.

nik9000 replied:
In the rest of Elasticsearch we implement enum serialization by hand rather than by ordinal. Well, we've mostly been doing that. I think SQL forked from a time when we were doing it by ordinal.

We transitioned because of cases like this - the "right" way to solve this is to send the key for `UNORDERED` to older node versions on the wire. If the serialization code was explicit then you'd be able to do it right in the enum itself.

nik9000 replied:
I'm fine pulling this out of the fix if that makes things easier to land. We can get it in a followup.

alex-spies replied:
All good, that's easy to get right - just missed it initially :)

astefan reviewed (changes requested):
One aspect that needs a deeper investigation regarding the order of the enum values.
Another aspect regarding adding a new csv index with one document just for one test.

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into fix-mv-dedupe

alex-spies committed:
Fix MvOrdering serialization

alex-spies committed:
Add is_rehired.keyword to employees.csv

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/mapping-multivalues.json:
```diff
@@ -0,0 +1,19 @@
+{
```
I realized that this requires larger changes since `employees.csv` does not seem to have doubles with duplicate multivalues. I'll go with a yaml test instead.

alex-spies committed:
Port added csv test to yaml test instead

alex-spies committed:
Add SORTED_ASCENDING to mv function tests

alex-spies committed:
Revert "Add is_rehired.keyword to employees.csv"

This reverts commit b8ba0ee7ee6b8f0ab55e17a35133c34c06a033e4.

astefan reviewed (approved):
LGTM

alex-spies commented:
Thanks for the reviews @nik9000 and @astefan !

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.14  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 107577`

alex-spies commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.14|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/107850">](https://github.com/elastic/elasticsearch/pull/107850)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->
